### DISCUSSION TITLE
Multiple Listeners: Why not multiple axum::serve ?

### DISCUSSION THREAD

**Discussion by moranbw** [tokio-rs/axum#2949] [Q&A] [answered]
### Summary

Mostly self explanatory from title...I was looking at the example for spawning multiple listeners here: https://github.com/tokio-rs/axum/blob/main/examples/listen-multiple-addrs/src/main.rs

This approach seems to drop down to using `hyper_util`.  But why not just spawn multiple instances of `axum::serve`?

This seems like an easier to understand pattern...but perhaps there are performance tradeoffs?

### axum version

0.7.7

**Comment by jplatte** [comment]
That's a very good question. The discussion at #1063 that led to the example being created doesn't really answer it 😅

Back then with `hyper::Server` it _looked_ like it could be more optimal to not spawn separate accept loop tasks, now I'm very skeptical. It might even be better to have two separate `axum::serve` tasks.

I'd be inclined to just delete the example. @mladedav thoughts?

**Reply by mladedav** [reply, marked as answer]
In an ideal world, I think it would be better to have it in one `serve` just so you can't add one later to one `serve` and forget to add it for althe other one. Unless that's what you want of course.

But I don't see practical issues with multiple concurrent `serve` calls. Well one that I can think of is that if a user fails to set non-blocking mode on ther TCP listeners then they'll be surprised that one of the servers doesn't seem to work at all. With single listener they'll blocking but it should just work by accident.

I think it should be just fine to get rid of the example since now it's needlessly complicated and the way to do this soon-ish will probably be to implement a `Listener` in 0.8 to achieve the same thing.

**Reply by moranbw** [reply]
@mladedav isn't  `tokio::net::TcpListener` always non-blocking? Would this concern only apply if you are using `std::net::TcpListener`?

**Reply by jplatte** [reply]
It's unfortunately not, see [`TcpListener::from_std`](https://docs.rs/tokio/latest/tokio/net/struct.TcpListener.html#method.from_std). But I don't think we really need to concern ourselves with that, it's not an axum-specific problem in any way.

Really the question is then: Should we update the example once #2941 is merged, or delete it instead. Even with a custom `Listener`, I'm not sure it's something we should promote. Or is there a reason you would consider that better, @mladedav? (excluding buggy user code)

**Reply by moranbw** [reply]
Sorry -- don't want to sidetrack from the actual conversation developing here...but @jplatte you again seemed to reference a TcpListener that is coming from `std::net::TcpListener`?

What if you are just creating a `tokio::net::TcpListener` like this, is this blocking or no: 
```rust
use tokio::net::TcpListener;

use std::io;

#[tokio::main]
async fn main() -> io::Result<()> {
    let listener = TcpListener::bind("127.0.0.1:2345").await?;

    // use the listener

    Ok(())
}
```

Sorry if this is a naive question.

**Reply by mladedav** [reply]
Nope, buggy user code is the only concern I have. We can delete it.

**Reply by mladedav** [reply]
@moranbw your example is non blocking, it would be just fine. The issue really is when you have something like

```rust
async fn run_server(listener: std::net::TcpListener) -> impl Future {
    // forgot to call listener.set_nonblocking(true);
    let listener = tokio::net::TcpListener::from_std(listener);
    axum::serve(listener, get_router())
}
```

It's not something you probably should do anyway, but for some reasons I do have code like this and I did see bugs where someone forgot to set one socket to non blocking in a server that used two sockets for two different things.

**Reply by moranbw** [reply]
Thanks for the info, @jplatte and @mladedav 🙏
